contributor-docs/adrs/adr-021-automated-prop-docs.md

```diff
@@ -0,0 +1,277 @@
+- Not well-documented and hard to debug
+- There are no existing plugins to help document React components
+- Seems like the point of this tool is to generate a static site for documentation. It's less useful for extracting type data and doing something custom with it.
+- I'm not sure how to print the returned statements/expressions (in the format of a TypeScript Symbol) in a way that will make it easier to debug. `JSON.stringify()` fails due to circular references.
+
+## Impact
+
+- We don't need to manually create `{ComponentName}.docs.json` for each component files for each component
+- Component metadata such as prop types and their descriptions will always be in sync with what's in the source code
+  - We only use this data for component docs on primer.style right now ([DataTable example](https://primer.style/components/data-table/react/draft#props)), but the possibilities are endless. This may even be useful for keeping Figma components in sync.
+- More thorough JSDoc comments provide more context when:
+  - working directly in PRC
+  - using PRC imports in an IDE
+
+## Decision
+
+TBD:
+- Will we move forward with automating prop docs?
+- If so, how? What are the suggested tools and process?
+
+### Potential paths forward
+
+- Move forward with `react-typescript-docgen` (preferred)
+  - Could be handled by @mperrotti with minimal help from an engineer
+- Move forward with TypeDoc
+  - Would require more time from one or more engineers
+- Build our own tooling to extract API data from component source code
+  - Would take a significant amount of high-effort work and would require more assistance from one or more engineer who are well-versed in TypeScript and ASTs
+- Manually maintain `.docs.json` files temporarily, figure out automation later
+  - If we're not comfortable with the previous directions, we could punt on this ADR and spend time updating `.docs.json` files for accuracy
+- Continue to manually maintain `.docs.json` files long-term
+  - Keep handling component API docs as we do today, but update this ADR to document how and why we document components this way
+
+
+## Open questions
+
+- Do we want to generate docs for internal component props? If so, where would they be rendered?
+- What should we do about hooks? It would be useful to have their data for docs, but it might be strange to keep that data in a file named **components**.json
+- What do we want to do about components with passthrough props? For example: [Button](https://github.com/primer/react/blob/219868e65214d6667c855c22d5c586b2f6b3348d/packages/react/src/Button/types.ts#L50)
+- Do we want to build `components.json` in CI the same way we do now? Also, it is explictly `gitignore`d in `packages/react/.gitignore`. For reference, here is the [step in our CI configuration](https://github.com/primer/react/blob/8629a4beade4821cf92da231a40a8b5106458680/.github/workflows/ci.yml#L254).
```

Comment:
Happy to go either way on this (either check in generated artifacts or not). I bet we'll still want a CI task for this for validation/make sure things can build just as a sanity check.

With generated files, I think the more "rolled up" they are the less valuable they are for the contributor. Meaning a huge generated file is harder to work with than one that is generated closer to the component. I think if we are dynamically generating `docs.json` files close to a component that could be helpful for a contributor but if it's rolling up into a broader `components.json` file it's probably fine to keep it as an artifact generated at build time.

Reply:
We could do something like this - generate a `.docs.json` file for each component that gets rolled up into a `components.json` as an artifact generated at build time.

Then we have to figure out what to do if we keep the `.docs.json` files around as fallbacks to provide missing info.